src/librbd/cache/SharedReadOnlyObjectDispatch.cc

```diff
@@ -0,0 +1,168 @@
+#include "librbd/io/Utils.h"
+#include "librbd/cache/SharedReadOnlyObjectDispatch.h"
+#include "osd/osd_types.h"
+#include "osdc/WritebackHandler.h"
+
+#include <vector>
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::cache::SharedReadOnlyObjectDispatch: " \
+                           << this << " " << __func__ << ": "
+
+namespace librbd {
+namespace cache {
+
+template <typename I, typename C>
+SharedReadOnlyObjectDispatch<I, C>::SharedReadOnlyObjectDispatch(
+    I* image_ctx) : m_image_ctx(image_ctx), m_cache_client(nullptr),
+    m_initialzed(false), m_object_store(nullptr) {
+}
+
+template <typename I, typename C>
+SharedReadOnlyObjectDispatch<I, C>::~SharedReadOnlyObjectDispatch() {
+    delete m_object_store;
+    delete m_cache_client;
+}
+
+// TODO if connect fails, init will return error to high layer.
+template <typename I, typename C>
+void SharedReadOnlyObjectDispatch<I, C>::init() {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 5) << dendl;
+
+  if (m_image_ctx->parent != nullptr) {
+    ldout(cct, 5) << "child image: skipping" << dendl;
+    return;
+  }
+
+  ldout(cct, 5) << "parent image: setup SRO cache client" << dendl;
+
+  std::string controller_path = ((CephContext*)cct)->_conf.get_val<std::string>("immutable_object_cache_sock");
+  if(m_cache_client == nullptr) {
+    m_cache_client = new C(controller_path.c_str(), m_image_ctx->cct);
+  }
+  m_cache_client->run();
+
+  int ret = m_cache_client->connect();
+  if (ret < 0) {
+    ldout(cct, 5) << "SRO cache client fail to connect with local controller: "
+                  << "please start ceph-immutable-object-cache daemon"
+		  << dendl;
+  } else {
+    ldout(cct, 5) << "SRO cache client to register volume "
+                  << "name = " << m_image_ctx->id
+                  << " on ceph-immutable-object-cache daemon"
+                  << dendl;
+
+    auto ctx = new FunctionContext([this](bool reg) {
+      handle_register_client(reg);
+    });
+
+    ret = m_cache_client->register_client(ctx);
+
+    if (ret >= 0) {
+      // add ourself to the IO object dispatcher chain
+      m_image_ctx->io_object_dispatcher->register_object_dispatch(this);
+      m_initialzed = true;
+    }
+  }
+}
+
+template <typename I, typename C>
+bool SharedReadOnlyObjectDispatch<I, C>::read(
+    const std::string &oid, uint64_t object_no, uint64_t object_off,
+    uint64_t object_len, librados::snap_t snap_id, int op_flags,
+    const ZTracer::Trace &parent_trace, ceph::bufferlist* read_data,
+    io::ExtentMap* extent_map, int* object_dispatch_flags,
+    io::DispatchResult* dispatch_result, Context** on_finish,
+    Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << "object_no=" << object_no << " " << object_off << "~"
+                 << object_len << dendl;
+
+  // if any failse, reads will go to rados
+  if(!m_cache_client->is_session_work() || m_cache_client == nullptr ||
```

Comment:
It would be nice if ```CacheClient``` would periodically try to re-connect to the daemon if it encounters a fault.

Reply:
good suggestions. i will implements this function which can support in-flight enable and re-connect.

Reply:
done. I have pushed one commit to implement re-connect and in-flight enable. thanks